controllers/posts.py

```
@@ -284,3 +304,85 @@ async def delete_post(token_payload: UserData, post_id: int) -> Response:
+
+
+# Endpoint: PATCH /posts/<post_id>/archive
+# Description: Archives a post in the database.
+# Parameters: post_id - ID of the post to update.
+# Authorization: patch:my-post or patch:any-post.
+@posts_endpoints.route("/posts/<post_id>/archive", methods=["PATCH"])
+@requires_auth(sah_config, ["patch:my-post", "patch:any-post"])
+async def archive_post(token_payload: UserData, post_id: int) -> Response:
+    archived = await _toggle_archive_post(
+        user_id=token_payload["id"],
+        post_id=post_id,
+        permissions=token_payload["role"]["permissions"],
+        method="archive",
+    )
+
+    return jsonify({"success": True, "archived": archived})
+
+
+# Endpoint: PATCH /posts/<post_id>/unarchive
+# Description: Unarchives a post in the database.
+# Parameters: post_id - ID of the post to update.
+# Authorization: patch:my-post or patch:any-post.
+@posts_endpoints.route("/posts/<post_id>/unarchive", methods=["PATCH"])
+@requires_auth(sah_config, ["patch:my-post", "patch:any-post"])
+async def unarchive_post(token_payload: UserData, post_id: int) -> Response:
+    unarchived = await _toggle_archive_post(
+        user_id=token_payload["id"],
+        post_id=post_id,
+        permissions=token_payload["role"]["permissions"],
+        method="unarchive",
+    )
+
+    return jsonify({"success": True, "unarchived": unarchived})
+
+
+async def _toggle_archive_post(
+    user_id: int,
+    post_id: int,
+    permissions: list[str],
+    method: Literal["archive", "unarchive"],
+) -> dict[str, Any]:
+    # Check if the post ID isn't an integer; if it isn't, abort
+    validator.check_type(post_id, "Post ID")
+    original_post: Post = await sah_config.db.one_or_404(
+        item_id=int(post_id),
+        item_type=Post,
+    )
+    if (method == "archive" and original_post.archived) or (
+        method == "unarchive" and not original_post.archived
+    ):
+        raise AuthError(
+            {
+                "code": 403,
```

I think 409 is more appropriate here. Although I'm not entirely sure whether we should throw an error here or just do nothing. Not set on it either way though

---

done